portal-kernel/test/unit/com/liferay/portal/kernel/url/URLBuilderTest.java

```diff
@@ -0,0 +1,231 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.portal.kernel.url;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.util.Portal;
+import com.liferay.portal.kernel.util.PortalUtil;
+import com.liferay.portal.kernel.util.StringBundler;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+
+import org.mockito.Mock;
+import org.mockito.Mockito;
+import org.mockito.MockitoAnnotations;
+
+/**
+ * @author Julius Lee
+ */
+public class URLBuilderTest {
+
+	@Before
+	public void setUp() {
+		MockitoAnnotations.initMocks(this);
+
+		Mockito.when(
+			_portal.stripURLAnchor(Mockito.anyString(), Mockito.anyString())
+		).thenAnswer(
+			input -> _testStripURLAnchor(
+				(String)input.getArguments()[0],
+				(String)input.getArguments()[1])
+		);
+
+		PortalUtil portalUtil = new PortalUtil();
+
+		portalUtil.setPortal(_portal);
+	}
+
+	@Test
+	public void testAddParameter() {
+		URLBuilder urlBuilder = URLBuilder.create(
+			"http://test.com"
+		).addParameter(
+			"testKey", "testValue"
+		);
+
+		Assert.assertEquals(
+			"http://test.com?testKey=testValue", urlBuilder.build());
+
+		urlBuilder.addParameter("testKey", "testValue");
+
+		Assert.assertEquals(
+			"http://test.com?testKey=testValue&testKey=testValue",
+			urlBuilder.build());
+	}
+
+	@Test
+	public void testAddParameterWithAnchor() {
+		URLBuilder urlBuilder = URLBuilder.create(
+			"http://test.com#TestAnchor"
+		).addParameter(
+			"testKey", "testValue"
+		);
+
+		Assert.assertEquals(
+			"http://test.com?testKey=testValue#TestAnchor", urlBuilder.build());
+
+		urlBuilder.addParameter(
+			"testKey", "testValue"
+		).build();
+
+		Assert.assertEquals(
+			"http://test.com?testKey=testValue&testKey=testValue#TestAnchor",
+			urlBuilder.build());
+	}
+
+	@Test
+	public void testAddParameterWithNullInput() {
+		Assert.assertNull(
+			URLBuilder.create(
+				null
+			).addParameter(
+				null, null
+			).setParameter(
+				null, null
+			).build());
+
+		// TODO: addParameter should not add if key is null
```

Comment:
@julschong @tinatian is this todo done yet? Either finish it or if it is already done, remove the comment here.

Reply:
@shuyangzhou 
Hi Shuyang,
the next 2 tasks for this petra jar is:
1. Remove the petra module completely since we just cleaned up URIBuilder using 3rd party library
2. Then, to move the param util logic from HttpComponentsUtil into URLBuilder completely. And remove the methods in HttpComponentsUtil

I plan to complete this TODO in step 2 above.
Let me know if this is ok

Reply:
Ok, then at least remove the comment, you can book keeping it somewhere locally. Don't keep it in the comment.